src/main.cpp

```diff
@@ -2522,7 +2563,50 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
     blockundo.vtxundo.reserve(block.vtx.size() - 1);
     int nChecked = 0;
     int nOrphansChecked = 0;
-    LOCK(cs_xval);
+    int nStartingHeight = chainActive.Height();
+
+    // Create a vector for storing hashes that will be deleted from the unverified and perverified txn sets.
+    // We will delete these hashes only if and when this block is the one that is accepted saving us the unnecessary 
+    // repeated locking and unlocking of cs_xval.
+    std::vector<uint256> vHashesToDelete;
+
+    // Create a temporary view of the UTXO set
+    CCoinsViewCache viewTempCache(pcoinsTip);
+
+    // Section for boost scoped lock on the scriptcheck_mutex
+    {
+
+    // Get the next available mutex and then use it to find the associated scriptcheckqueue. Then lock this thread
+    // with the mutex so other thread can used this scriptcheckqueue until all inputs have been checked and the 
+    // scriptcheckqueue threads have all returned.
+    boost::shared_ptr<boost::mutex> scriptcheck_mutex = allScriptCheckQueues.GetScriptCheckMutex();
+    if (fParallel) cs_main.unlock(); // unlock cs_main, we may be waiting here for a while before aquiring the scoped lock below
+    boost::mutex::scoped_lock lock(*scriptcheck_mutex);
```

Comment:
It works fine...take a look a the newer code.  Much has changed and been streamlined.  Mostly everything is now in parallel.h and parallel.cpp.   Any code is "brittle" and requires the coder to understand the code he's dealing with.  The problem we have here is we can either support two sets of code for ConnectBlock(), one that is for parallel validation and one without, but that will end up being much worse IMO for potential code breakage, if/when developers make changes to one without making them to the other.  Take a look at the newer code and then comment again.